sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -454,6 +492,322 @@ pi_result OCL(piEnqueueMemBufferMap)(
+                              pi_uint32 alignment) {
+
+  void *Ptr = nullptr;
+  pi_result RetVal = PI_INVALID_OPERATION;
+
+  // First we need to look up the function pointer
+  // It would be good if we could store it for future reuse
+  // Are statics ok?
+  clHostMemAllocINTEL_fn FuncPtr;
+  RetVal = getExtFuncFromContext<clHostMemAllocINTEL_fn>(
+      context, "clHostMemAllocINTEL", &FuncPtr);
+
+  if (RetVal == PI_SUCCESS) {
+    Ptr = FuncPtr(cast<cl_context>(context),
+                  cast<cl_mem_properties_intel *>(properties), size, alignment,
+                  cast<cl_int *>(&RetVal));
+  }
+
+  *result_ptr = Ptr;
+
+  return RetVal;
+}
+
+// Allocates device memory.
+pi_result OCL(piDeviceMemAlloc)(void **result_ptr, pi_context context,
+                                pi_device device,
+                                pi_usm_mem_properties *properties, size_t size,
+                                pi_uint32 alignment) {
+
+  void *Ptr = nullptr;
+  pi_result RetVal = PI_INVALID_OPERATION;
+
+  // First we need to look up the function pointer
+  // It would be good if we could store it for future reuse
+  // Are statics ok?
+  clDeviceMemAllocINTEL_fn FuncPtr;
+  RetVal = getExtFuncFromContext<clDeviceMemAllocINTEL_fn>(
+      context, "clDeviceMemAllocINTEL", &FuncPtr);
+
+  if (RetVal == PI_SUCCESS) {
+    Ptr = FuncPtr(cast<cl_context>(context), cast<cl_device_id>(device),
+                  cast<cl_mem_properties_intel *>(properties), size, alignment,
+                  cast<cl_int *>(&RetVal));
+  }
+
+  *result_ptr = Ptr;
+
+  return RetVal;
+}
+
+// Allocates memory accessible on host and device.
+pi_result OCL(piSharedMemAlloc)(void **result_ptr, pi_context context,
+                                pi_device device,
+                                pi_usm_mem_properties *properties, size_t size,
+                                pi_uint32 alignment) {
+
+  void *Ptr = nullptr;
+  pi_result RetVal = PI_INVALID_OPERATION;
+
+  // First we need to look up the function pointer
+  // It would be good if we could store it for future reuse
+  // Are statics ok?
+  clSharedMemAllocINTEL_fn FuncPtr;
+  RetVal = getExtFuncFromContext<clSharedMemAllocINTEL_fn>(
+      context, "clSharedMemAllocINTEL", &FuncPtr);
+
+  if (RetVal == PI_SUCCESS) {
+    Ptr = FuncPtr(cast<cl_context>(context), cast<cl_device_id>(device),
+                  cast<cl_mem_properties_intel *>(properties), size, alignment,
+                  cast<cl_int *>(&RetVal));
+  }
+
+  *result_ptr = Ptr;
+
+  return RetVal;
+}
+
+pi_result OCL(piMemFree)(pi_context context, void *ptr) {
+
+  clMemFreeINTEL_fn FuncPtr;
+  pi_result RetVal = getExtFuncFromContext<clMemFreeINTEL_fn>(
+      context, "clMemFreeINTEL", &FuncPtr);
+
+  if (RetVal != PI_SUCCESS)
+    return RetVal;
+
+  return cast<pi_result>(FuncPtr(cast<cl_context>(context), ptr));
+}
+
+// Sets up pointer arguments to CL kernels
```

Comment:
How is this exactly different from the regular piKernelSetArg? Is this doing extra indirection? Why?

Reply:
Yes. It's doing extra indirection b/c OpenCL, in all its infinite wisdom, decided that pointer args should be set by value instead of by address + size.

I think the OpenCL runtime might also be doing something special on its end, not sure.